src/lib/pages/getStaticProps/setup.js

```diff
@@ -1,51 +1,56 @@
 const isRouteWithFallback = require('../../helpers/isRouteWithFallback')
-const { logTitle, logItem } = require('../../helpers/logger')
-const setupNetlifyFunctionForPage = require('../../helpers/setupNetlifyFunctionForPage')
-const setupStaticFileForPage = require('../../helpers/setupStaticFileForPage')
+const { logTitle } = require('../../helpers/logger')
 
 const getPages = require('./pages')
 
 // Copy pre-rendered SSG pages
 const setup = async ({ functionsPath, publishPath }) => {
   logTitle('🔥 Copying pre-rendered pages with getStaticProps and JSON data to', publishPath)
-
   // Keep track of the functions that have been set up, so that we do not set up
   // a function for the same file path twice
-  const filePathsDone = []
-
+  const filePathsDone = new Set()
   const pages = await getPages()
 
-  await asyncForEach(pages, async ({ route, dataRoute, srcRoute }) => {
-    logItem(route)
-
-    // Copy pre-rendered HTML page
-    const htmlPath = getFilePathForRoute(route, 'html')
-    await setupStaticFileForPage({ inputPath: htmlPath, publishPath })
-
-    // Copy page's JSON data
-    const jsonPath = getFilePathForRoute(route, 'json')
-    await setupStaticFileForPage({
-      inputPath: jsonPath,
-      outputPath: dataRoute,
-      publishPath,
-    })
-
-    // Set up the Netlify function (this is ONLY for preview mode)
-    const relativePath = getFilePathForRoute(srcRoute || route, 'js')
-    const filePath = join('pages', relativePath)
-
-    // Skip if we have already set up a function for this file
-    // or if the source route has a fallback (handled by getStaticPropsWithFallback)
-    if (filePathsDone.includes(filePath) || (await isRouteWithFallback(srcRoute))) return
-
-    logItem(filePath)
-    await setupNetlifyFunctionForPage({ filePath, functionsPath })
-    filePathsDone.push(filePath)
-  })
+  const jobs = []
+
+  await Promise.all(
+    pages.map(async ({ route, dataRoute, srcRoute }) => {
+      // Copy pre-rendered HTML page
+      const htmlPath = getFilePathForRoute(route, 'html')
+
+      jobs.push({ type: 'static', inputPath: htmlPath, publishPath })
+
+      // Copy page's JSON data
+      const jsonPath = getFilePathForRoute(route, 'json')
+      jobs.push({
+        type: 'static',
+        inputPath: jsonPath,
+        outputPath: dataRoute,
+        publishPath,
+      })
+
+      // Set up the Netlify function (this is ONLY for preview mode)
+      const relativePath = getFilePathForRoute(srcRoute || route, 'js')
+      const filePath = join('pages', relativePath)
+
+      // Skip if we have already set up a function for this file
+
+      if (filePathsDone.has(filePath)) {
+        return
+      }
+      filePathsDone.add(filePath)
+
+      // or if the source route has a fallback (handled by getStaticPropsWithFallback)
+      if (await isRouteWithFallback(srcRoute)) {
+        return
+      }
+      jobs.push({ type: 'function', filePath, functionsPath })
+    }),
+  )
+  return jobs
 }
```

Comment:
The callback in the map is async (because it has to await `isRouteWithFallback`) , so it automatically returns a promise which resolves (to `void`) when it's complete. This means that the map returns an array of promises (one for each page). Passing these to an `await`ed `Promise.all` means that it waits until all of the functions are complete. With an async callback like that, the map will _immediately_ return the array of promises, before the callback has executed for each of the elements. It would be somehting like this. Apologies if you know all this:

```js
const result = pages.map(async () => {
await something()
})

// We get here immediately, before the callbacks execute
console.log(result)
// [ <unresolved Promise>, <unresolved Promise>, <unresolved Promise>, ...]

// ...the callbacks execute in parallel, and resolve one by one. we might wait here a while
const done = await Promise.all(result)
// once we're here, the jobs are all complete
console.log(result)

// [ <resolved Promise>, <resolved Promise>, <resolved Promise>, ...]

console.log(done)

// [void, void, void, ...]

return
// 

```

The reason I'm using the jobs array from outside of the map is so that I can conditionally add elements. If I returned them, I'd end up with an array with `undefined` elements for the pages we're skipping. We could then filter that array, but that means we'd need to loop through them again. Pushing onto an array outside means we only need to loop through once. I could use `reduce` for the same effect, but I generally find that's harder to read.

Reply:
ooh ok i dont think i processed that the map callback is async. ok wow this is wild!!!